### PR TITLE
Added openstack engine

### PR DESCRIPTION
**Pull request by mikezuniga**
Added openstack engine
Added kubernetes heat template to use with opesntack engine
Added documentation on how to use the openstack engine

### CODE REVIEW

File: docs/kqueen.rst [hunk 1]
```diff
@@ -360,6 +360,31 @@ Provision a Kubernetes cluster
...
+4. Copy ``User Name``, ``Project Name``, ``Authentication URL``
+5. Navigate to ``Project`` -> ``Network`` -> ``Networking``
+6. Click on your Private Network, then click on a Subnet of your Private Network.
+7. Copy the ID of the Subnet and the Network ID
+8. Navigate to ``Project`` -> ``Network`` -> ``Networking``
+9. Click on your Public Network, then copy the public network ID
+10. Import into Glance the following image http://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1801-01.qcow2
+11. Download the following heat template https://github.com/Mirantis/kqueen/prod/openstack/heat-templates/kubernetes.yaml
+12. Under the ``parameters:`` section replace the private_net_id, private_subnet_id and public_net_id default values with ones you copied.
+13. Login in to KQueen web UI.
```
Comment (ekhomyakova) [inline docs/kqueen.rst]: Log in to

File: docs/kqueen.rst [hunk 2]
```diff
@@ -360,6 +360,31 @@ Provision a Kubernetes cluster
 
 You can provision a Kubernetes cluster using various community of engines, such as Google Kubernetes engine or Azure Kubernetes Service.
 
+To provision a Kubernetes cluster using the Openstack Engine:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Login into your Openstack Horizon Dashboard
```
Comment (ekhomyakova) [inline docs/kqueen.rst]: It's better to use `#. ` instead of step numbers for automatic numeration

File: docs/kqueen.rst [hunk 3]
```diff
@@ -360,6 +360,31 @@ Provision a Kubernetes cluster
...
+7. Copy the ID of the Subnet and the Network ID
+8. Navigate to ``Project`` -> ``Network`` -> ``Networking``
+9. Click on your Public Network, then copy the public network ID
+10. Import into Glance the following image http://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud-1801-01.qcow2
+11. Download the following heat template https://github.com/Mirantis/kqueen/prod/openstack/heat-templates/kubernetes.yaml
+12. Under the ``parameters:`` section replace the private_net_id, private_subnet_id and public_net_id default values with ones you copied.
+13. Login in to KQueen web UI.
+14. From the ``Create Provisioner`` tab, select ``Openstack Engine``.
+15. Fill the form with the User Name, Password, Project/Tenant Name and Authentication URL
+16. Copy/Paste the modify kubernetes.yaml heat template under the field ``Heat template to use for building k8s clusters``
```
Comment (ekhomyakova) [inline docs/kqueen.rst]: > Copy/Paste the modify

Looks like a typo somewhere

File: prod/openstack/heat-templates/kubernetes.yaml [hunk 4]
```diff
@@ -0,0 +1,128 @@
+heat_template_version: 2013-05-23
+
+description: >
+  Heat Kubernetes template to support linux OS, using only Heat Openstack-native
+  resource types, and without the requirement for heat-cfntools in the image.
+
+parameters:
+
```
Comment (ekhomyakova) [inline prod/openstack/heat-templates/kubernetes.yaml]: Looks like this empty line is redundant

File: prod/openstack/heat-templates/kubernetes.yaml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    description: Name of a KeyPair to enable SSH access to the instance
+    default: secret
+  instance_type:
+    type: string
+    description: Instance type for k8s server
+    default: m1.small
+  image_id:
+    type: string
+    description: >
+      Image to use for kubernetes servers .
```
Comment (ekhomyakova) [inline prod/openstack/heat-templates/kubernetes.yaml]: Please remove the whitespace before the dot

File: kqueen/engines/openstack.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+                'label': 'Project/Tenant Name',
+                'order': 2,
+                'validators': {
+                    'required': True
+                }
+            },
+            'os_auth_url': {
+                'type': 'text',
+                'label': 'Authentication URL (keystone)',
+                'order': 6,
```
Comment (ekhomyakova) [inline kqueen/engines/openstack.py]: `'order': 3,` is better, I think

File: kqueen/engines/openstack.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+                'validators': {
+                    'required': True
+                }
+            },
+            'os_heat_k8s_template': {
+                'type': 'textarea',
+                'label': 'Heat template to use for building k8s clusters',
+                'order': 8,
```
Comment (ekhomyakova) [inline kqueen/engines/openstack.py]: `'order': 4,`

File: kqueen/engines/openstack.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        Implementation of :func:`~kqueen.engines.base.BaseEngine.provision`
+        """
+        try:
+            response = self.client.stacks.create(files={}, disable_rollback=True, name=self.name, template=self.os_heat_k8s_template)
+            self.cluster.id = response.id
+            self.cluster.save()
+            # TODO: check if provisioning response is healthy
+        except Exception as e:
+            msg = 'Creating cluster {} failed with following reason:'.format(self.cluster.id)
```
Comment (ekhomyakova) [inline kqueen/engines/openstack.py]: with the following

File: kqueen/engines/openstack.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+                'validators': {
+                    'required': True
+                }
+            },
+            'os_auth_url': {
+                'type': 'text',
+                'label': 'Authentication URL (keystone)',
+                'order': 6,
+                'validators': {
+                    'required': True
```
Comment (katyafervent) [inline kqueen/engines/openstack.py]: Please, add 'url': True to the validators dict

File: kqueen/engines/openstack.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                'default': 1,
+                'validators': {
+                    'required': True,
+                    'max': 1,
+                    'number': True
+                }
+            },
+            'name': {
+                'type': 'string',
+                'label': 'Cluster Name',
```
Comment (katyafervent) [inline kqueen/engines/openstack.py]: name is added automatically, you'll have two fields with names

File: kqueen/engines/openstack.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+                'default': 'k8s',
+                'validators': {
+                    'required': True
+                }
+            }
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (mikezuniga) [reply kqueen/engines/openstack.py]: Already fixed on the last commit

File: prod/openstack/heat-templates/kubernetes.yaml [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+            wget https://github.com/kubernetes/minikube/releases/download/v0.26.1/minikube-linux-amd64
+            mv minikube-linux-amd64 /bin/minikube
+            chmod 755 /bin/minikube
+            minikube start --vm-driver=none 
+            cat <<EOF > /var/www/html/kubeconfig
+            apiVersion: v1
+            clusters:
+            - cluster:
+                #certificate-authority: /root/.minikube/ca.crt
+                insecure-skip-tls-verify: true
```
Comment (epcim) [inline prod/openstack/heat-templates/kubernetes.yaml]: nice default

File: prod/openstack/heat-templates/kubernetes.yaml [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            yum install -y wget curl socat httpd
+            yum install -y yum-utils device-mapper-persistent-data lvm2
+            yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
+            yum install -y docker-ce
+            service docker start
+            service httpd start
+            cat <<EOF > /etc/yum.repos.d/kubernetes.repo
+            [kubernetes]
+            name=Kubernetes
+            baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64
```
Comment (epcim) [inline prod/openstack/heat-templates/kubernetes.yaml]: will not work at offline/restricted deployments, the baseurl must be configurable.

**Comment by naumvd95** [comment]
@mikezuniga do we still need that PR?

**Pull request closed without merging**